bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCallbackUtils.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.bookkeeper.client;
+
+import com.google.common.collect.Iterators;
+import java.util.Enumeration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import static org.apache.bookkeeper.client.LedgerHandle.LOG;
+import org.apache.bookkeeper.client.api.Handle;
+import org.apache.bookkeeper.client.api.ReadHandle;
+
+/**
+ * Utility for callbacks
+ *
+ */
+class SyncCallbackUtils {
+
+    /**
+     * Wait for a result. This is convenience method to implement callbacks
+     *
+     * @param <T>
+     * @param future
+     * @return
+     * @throws InterruptedException
+     * @throws BKException
+     */
+    public static <T> T waitForResult(CompletableFuture<T> future) throws InterruptedException, BKException {
+        try {
+            return future.get();
+        } catch (ExecutionException err) {
+            if (err.getCause() instanceof BKException) {
+                throw (BKException) err.getCause();
+            } else {
+                BKException unexpectedConditionException
+                    = BKException.create(BKException.Code.UnexpectedConditionException);
+                unexpectedConditionException.initCause(err.getCause());
+                throw unexpectedConditionException;
+            }
+
+        }
+    }
+
+    /**
+     * Handle the Response Code and transform it to a BKException
+     *
+     * @param <T>
+     * @param rc
+     * @param result
+     * @param future
+     */
+    public static <T> void finish(int rc, T result, CompletableFuture<T> future) {
+        if (rc != BKException.Code.OK) {
+            future.completeExceptionally(BKException.create(rc).fillInStackTrace());
+        } else {
+            future.complete(result);
+        }
+    }
+
+    static class SyncCreateCallback implements AsyncCallback.CreateCallback {
+
+        private final CompletableFuture future;
```

Comment:
I think I have a few comments about this? I'd prefer enforcing type here.

Reply:
done, I found a trick to make it work with <? super LedgeHandle> but I needed to add "implements WriteAdvHandle" in LedgerHandle

Reply:
if LedgerHandleAdv and LedgerHandle are going to be different, why can't have two different callbacks?

SyncCreateCallback => CompletableFuture<? super LedgerHandle>
SyncCreateAdvCallback => CompletableFuture<? super LedgerHandleAdv>

rather than doing the trick for LedgerHandle to implement WriteAdvHandle.

Reply:
Good idea, I will do as you are suggesting